From 954d5f522ba9a85f85d75e213450fdf395326f81 Mon Sep 17 00:00:00 2001
From: xulei <stone.xulei@huawei.com>
Date: Thu, 30 Dec 2021 16:25:51 +0800
Subject: [PATCH] kvm: SMM: fix losing SMI problem
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 954d5f522ba9a85f85d75e213450fdf395326f81


virt inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I4NP0K
CVE: NA
-------------------

fix losing SMI problem

Signed-off-by: xulei <stone.xulei@huawei.com>
Signed-off-by: Jingyi Wang <wangjingyi11@huawei.com>
Reviewed-by: Zenghui Yu <yuzenghui@huawei.com>
Reviewed-by: Wei Li <liwei391@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/x86/kvm/x86.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 12db47c8bd3f..e33414f36dba 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -4227,6 +4227,11 @@ static int kvm_vcpu_ioctl_nmi(struct kvm_vcpu *vcpu)
 
 static int kvm_vcpu_ioctl_smi(struct kvm_vcpu *vcpu)
 {
+	if (is_smm(vcpu)) {
+		vcpu->arch.hflags &= ~HF_SMM_MASK;
+		vcpu->arch.smi_pending = 0;
+	}
+
 	kvm_make_request(KVM_REQ_SMI, vcpu);
 
 	return 0;
-- 
2.26.2

